pkg/operator/controllers/banner/banner_controller_test.go

```diff
@@ -0,0 +1,182 @@
+package banner
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+
+	consolev1 "github.com/openshift/api/console/v1"
+	consolefake "github.com/openshift/client-go/console/clientset/versioned/fake"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+	arofake "github.com/Azure/ARO-RP/pkg/operator/clientset/versioned/fake"
+	utillog "github.com/Azure/ARO-RP/pkg/util/log"
+)
+
+func TestBannerReconcile(t *testing.T) {
+	r := Reconciler{log: utillog.GetLogger()}
+	for _, tt := range []struct {
+		name            string
+		oldCN           consolev1.ConsoleNotification
+		bannerSetting   string
+		expectBanner    bool
+		expectedMessage string
+		expectedErr     error
+		featureFlag     bool
+	}{
+		{
+			name:          "Wrong banner setting",
+			bannerSetting: "WRONG",
+			expectBanner:  false,
+			expectedErr:   fmt.Errorf("wrong banner setting '%s'", "WRONG"),
+			featureFlag:   true,
+		},
+		{
+			name:          "No banner when feature disabled",
+			bannerSetting: "ContactSupport",
+			expectBanner:  false,
+			featureFlag:   false,
+		},
+		{
+			name: "Banner not deleted when feature disabled",
+			oldCN: consolev1.ConsoleNotification{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: BannerName,
+				},
+				Spec: consolev1.ConsoleNotificationSpec{
+					Text:            "OLD BANNER TEXT",
+					Location:        consolev1.BannerTop,
+					Color:           "#000",
+					BackgroundColor: "#ff0",
+				},
+			},
+			bannerSetting:   "",
+			expectBanner:    true,
+			expectedMessage: "OLD BANNER TEXT",
+			featureFlag:     false,
+		},
```

Comment:
Yes, we should :)

Reply:
I would expect the code to be completely turned off when the feature flag is off. 

When I turn it off, it's off, not trying to clean up its objects because I might want that object to thave a manual override and stay behind - not get reconciled to nil. I know this use case doesn't apply here (1. we don't need manual overrides on a pop up text in a console, 2. if the need ever arises we could use an ConsoleNotification with a different name), but that's the general behaviour I expect from all reconcilers behind a feature: false means the code is not interfering with anything.

Reply:
We discussed it briefly on a call. The idea is that we use feature flag to disable reconciliation, not to disable the banner. It is just a mechanism which we use to turn off the controller if it is broken.

So basically:

* To disable the banner - set content to `BannerDisabled`.
* To disable reconciler - set `ReconcileBanner` to `false`

If we start using feature flag to clean up the banner then:

* We end up with 2 ways to disable the banner (`BannerDisabled` content and `ReconcileBanner` set to `false`)
* Feature flag no longer serves the porpuse of an emergency brake: we still need to run reconciliation and this reconciliation might have a a bug on it's own so we are screwwed and need to deploy a new operator again.

Reply:
Also I think this goes into a wider discussion on what we mean by a feature flag in ARO Operator in general. I'm adding a design/definition doc and Architecture call topic to discuss it.